include/circt/Scheduling/Problems.h

```diff
@@ -454,6 +454,29 @@ class ModuloProblem : public virtual CyclicProblem,
   virtual LogicalResult verify() override;
 };
 
+/// This class models the accumulation of physical propagation delays on
+/// combinational paths along SSA dependences on a cyclic scheduling problem.
+///
+/// Each operator type is annotated with estimated values for incoming and
+/// outgoing delays. Combinational operators (zero-latency, no internal
+/// registers) have only a single delay; this important special case is modeled
+/// by setting the incoming and outgoing delays to the same values.
+///
+/// A solution to this problem comprises per-operation start times in a
+/// continuous unit, e.g. in nanoseconds, inside the discrete time steps/cycles
+/// determined by the underlying scheduling problem. Its solution can be used to
+/// construct a pipelined datapath with a fixed, integer initiation interval,
+/// in which the execution of multiple iterations/samples/etc. may overlap.
+class ChainingCyclicProblem : public virtual ChainingProblem,
+                              public virtual CyclicProblem {
+  DEFINE_COMMON_MEMBERS(ChainingCyclicProblem)
```

Comment:
Do not do things like this.  This can be solved with inheritance, for example inheriting from `Problem<ChainingCyclicProblem>`.

Reply:
I don't understand how you want me to declare the class `ChainingCyclicProblem`. As expected, when I try to inherit from `Problem<ChainingCyclicProblem>`, the compilation fails with the error `unknown template name 'Problem'`. How should I change this part of the code?

Reply:
I think for the sake of this PR, we should keep the macro solution, because that's how all other problem classes in the infra are defined. I can a file an issue later to rethink which inheritance mechanism is best-suited for the scheduling infra. @darthscsi OK?

Reply:
I filed #6517 for this.